pkg/services/folder/folderimpl/folder_unifiedstorage.go

```diff
@@ -0,0 +1,698 @@
+	"github.com/grafana/grafana/pkg/services/apiserver/endpoints/request"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	"github.com/grafana/grafana/pkg/services/dashboards/dashboardaccess"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/folder"
+	"github.com/grafana/grafana/pkg/services/guardian"
+	"github.com/grafana/grafana/pkg/services/store/entity"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/util"
+)
+
+// interface to allow for testing
+type folderK8sHandler interface {
+	getClient(ctx context.Context, orgID int64) (dynamic.ResourceInterface, bool)
+	getNamespace(orgID int64) string
+}
+
+var _ folderK8sHandler = (*foldk8sHandler)(nil)
+
+type foldk8sHandler struct {
+	cfg        *setting.Cfg
+	namespacer request.NamespaceMapper
+	gvr        schema.GroupVersionResource
+}
+
+func (s *Service) getFoldersFromApiServer(ctx context.Context, q folder.GetFoldersQuery) ([]*folder.Folder, error) {
+	if q.SignedInUser == nil {
+		return nil, folder.ErrBadRequest.Errorf("missing signed in user")
+	}
+
+	qry := folder.NewGetFoldersQuery(q)
+	permissions := q.SignedInUser.GetPermissions()
+	folderPermissions := permissions[dashboards.ActionFoldersRead]
+	qry.AncestorUIDs = make([]string, 0, len(folderPermissions))
+	if len(folderPermissions) == 0 && !q.SignedInUser.GetIsGrafanaAdmin() {
+		return nil, nil
+	}
+	for _, p := range folderPermissions {
+		if p == dashboards.ScopeFoldersAll {
+			// no need to query for folders with permissions
+			// the user has permission to access all folders
+			qry.AncestorUIDs = nil
+			break
+		}
+		if folderUid, found := strings.CutPrefix(p, dashboards.ScopeFoldersPrefix); found {
+			if !slices.Contains(qry.AncestorUIDs, folderUid) {
+				qry.AncestorUIDs = append(qry.AncestorUIDs, folderUid)
+			}
+		}
+	}
+
+	var dashFolders []*folder.Folder
+	var err error
+
+	ctx = identity.WithRequester(ctx, q.SignedInUser)
+	dashFolders, err = s.unifiedStore.GetFolders(ctx, qry)
+	if err != nil {
+		return nil, folder.ErrInternal.Errorf("failed to fetch subfolders: %w", err)
+	}
+
+	return dashFolders, nil
+}
+
+func (s *Service) getFromApiServer(ctx context.Context, q *folder.GetFolderQuery) (*folder.Folder, error) {
+	if q.SignedInUser == nil {
+		return nil, folder.ErrBadRequest.Errorf("missing signed in user")
+	}
+
+	if q.UID != nil && *q.UID == accesscontrol.GeneralFolderUID {
+		return folder.RootFolder, nil
+	}
+
+	if q.UID != nil && *q.UID == folder.SharedWithMeFolderUID {
+		return folder.SharedWithMeFolder.WithURL(), nil
+	}
+
+	var dashFolder *folder.Folder
+	var err error
+	switch {
+	case q.UID != nil:
+		if *q.UID == "" {
+			return &folder.GeneralFolder, nil
+		}
+		dashFolder, err = s.unifiedStore.Get(ctx, *q)
+		if err != nil {
+			return nil, err
+		}
+	// nolint:staticcheck
+	case q.ID != nil:
+		// not implemented
+	case q.Title != nil:
+		// not implemented
```

Comment:
Are these safe to not be implemented?

Reply:
I am planning to have these powered by the query/indexing server on a follow-up PR.

Those are the parts that will break:

https://github.com/grafana/grafana/blob/0e953a8ff379364a5c29bb8b760a9985f4e99eee/pkg/services/provisioning/alerting/rules_provisioner.go#L121

https://github.com/grafana/grafana/blob/0e953a8ff379364a5c29bb8b760a9985f4e99eee/pkg/services/dashboardimport/service/service.go#L48

https://github.com/grafana/grafana/blob/0e953a8ff379364a5c29bb8b760a9985f4e99eee/pkg/api/folder.go#L193

What do you think, it would be worth to take the oportunity and implement them on this PR?

Reply:
Totally fine to keep as a separate PR!